File: libs/python/src/payment_platform/telemetry.py
"""
OpenTelemetry tracing setup for Python services.

Uses OTLP gRPC exporter to send spans to the collector.
Auto-instruments FastAPI with standard HTTP spans and W3C trace context propagation.
"""

import logging
from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION, SERVICE_NAMESPACE, DEPLOYMENT_ENVIRONMENT
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.trace.sampling import ALWAYS_ON

_logger = logging.getLogger(__name__)


def setup_telemetry(
    app: FastAPI,
    service_name: str,
    service_version: str,
    exporter_endpoint: str,
    environment: str = "local",
) -> TracerProvider:
    """
    Initialize OpenTelemetry SDK with OTLP gRPC exporter and FastAPI auto-instrumentation.

    Args:
        app: FastAPI application instance
        service_name: Name for this service (e.g., "fraud-service")
        service_version: Version string (e.g., "0.1.0")
        exporter_endpoint: OTLP gRPC endpoint (e.g., "http://otel-collector:4317")
        environment: Deployment environment (e.g., "local", "staging", "production")

    Returns:
        TracerProvider — caller is responsible for calling shutdown() on exit.
    """
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
        SERVICE_NAMESPACE: "payment-api",
        DEPLOYMENT_ENVIRONMENT: environment,
    })

    # Strip http:// prefix for gRPC if present
    grpc_endpoint = exporter_endpoint
    if grpc_endpoint.startswith("http://"):
        grpc_endpoint = grpc_endpoint[len("http://"):]
    elif grpc_endpoint.startswith("https://"):
        grpc_endpoint = grpc_endpoint[len("https://"):]

    exporter = OTLPSpanExporter(endpoint=grpc_endpoint, insecure=True)
    processor = BatchSpanProcessor(exporter)

    provider = TracerProvider(
        resource=resource,
        sampler=ALWAYS_ON,
        active_span_processor=processor,
    )
    trace.set_tracer_provider(provider)

    # Auto-instrument FastAPI
    FastAPIInstrumentor.instrument_app(app)

    _logger.info("OTel tracing initialized: service=%s endpoint=%s", service_name, grpc_endpoint)
    return provider

File: libs/nodejs/src/telemetry.ts
/**
 * OpenTelemetry tracing setup for Node.js services.
 * Uses OTLP gRPC exporter to send spans to the collector.
 */

import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { FastifyInstrumentation } from '@opentelemetry/instrumentation-fastify';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { Resource } from '@opentelemetry/resources';
import {
  ATTR_SERVICE_NAME,
  ATTR_SERVICE_VERSION,
  ATTR_SERVICE_NAMESPACE,
  ATTR_DEPLOYMENT_ENVIRONMENT_NAME,
} from '@opentelemetry/semantic-conventions';

let sdk: NodeSDK | null = null;

/**
 * Initialize the OpenTelemetry SDK.
 * Must be called BEFORE creating the Fastify server.
 */
export function initTelemetry(
  serviceName: string,
  exporterEndpoint: string,
  serviceVersion: string = '0.1.0',
  environment: string = 'local',
): NodeSDK {
  const resource = new Resource({
    [ATTR_SERVICE_NAME]: serviceName,
    [ATTR_SERVICE_VERSION]: serviceVersion,
    [ATTR_SERVICE_NAMESPACE]: 'payment-api',
    [ATTR_DEPLOYMENT_ENVIRONMENT_NAME]: environment,
  });

  // Strip http:// prefix for gRPC
  let grpcEndpoint = exporterEndpoint;
  if (grpcEndpoint.startsWith('http://')) {
    grpcEndpoint = grpcEndpoint.slice(7);
  }

  const exporter = new OTLPTraceExporter({
    url: `http://${grpcEndpoint}`,
  });

  sdk = new NodeSDK({
    resource,
    traceExporter: exporter,
    instrumentations: [
      new HttpInstrumentation(),
      new FastifyInstrumentation(),
    ],
  });

  sdk.start();

  // Graceful shutdown
  process.on('SIGTERM', async () => {
    if (sdk) {
      await sdk.shutdown();
    }
  });

  console.log(`OTel tracing initialized: service=${serviceName} endpoint=${grpcEndpoint}`);
  return sdk;
}

/** Get the active SDK instance for manual shutdown. */
export function getSDK(): NodeSDK | null {
  return sdk;
}
